packages/runner/src/recipe-manager.ts

```diff
@@ -83,36 +82,67 @@ export class RecipeManager implements IRecipeManager {
       ? createRef({ src }, "recipe source").toString()
       : createRef(recipe, "recipe").toString();
 
+    this.recipeIdMap.set(generatedId, recipe as Recipe);
+    if (src) this.recipeProgramMap.set(recipe as Recipe, src);
+
     return generatedId;
   }
 
-  async registerRecipe(
+  saveRecipe(
     { recipeId, space, recipe, recipeMeta }: {
       recipeId: string;
       space: MemorySpace;
-      recipe: Recipe | Module;
-      recipeMeta: RecipeMeta;
+      recipe?: Recipe | Module;
+      recipeMeta?: RecipeMeta;
     },
-  ): Promise<boolean> {
-    if (!recipeMeta.src && !recipeMeta.program) {
-      return false;
-    }
-
+  ): boolean {
     // FIXME(ja): should we update the recipeMeta if it already exists? when does this happen?
     if (this.recipeMetaMap.has(recipe as Recipe)) {
       return true;
     }
 
-    const recipeMetaCell = await this.getRecipeMetaCell({ recipeId, space });
-    recipeMetaCell.set(recipeMeta);
-    await this.runtime.storage.syncCell(recipeMetaCell);
-    await this.runtime.storage.synced();
+    if (!recipe) {
+      recipe = this.recipeById(recipeId);
+      if (!recipe) {
+        throw new Error(`Recipe ${recipeId} not loaded`);
+      }
+    }
+
+    if (!recipeMeta) {
+      const src = this.recipeProgramMap.get(recipe as Recipe);
+      recipeMeta = {
+        id: recipeId,
+        src: typeof src === "string" ? src : undefined,
+        program: typeof src === "object" ? src : undefined,
+      };
+    }
+
+    if (!recipeMeta.src && !recipeMeta.program) {
+      return false;
+    }
 
-    this.recipeIdMap.set(recipeId, recipe as Recipe);
+    const recipeMetaCell = this.getRecipeMetaCell({ recipeId, space });
+    recipeMetaCell.set(recipeMeta);
     this.recipeMetaMap.set(recipe as Recipe, recipeMetaCell);
     return true;
   }
 
+  saveAndSyncRecipe(
```

Comment:
we don't check the return value, should we?  or should we simplify the implementation to not have all these resolves?

Reply:
removed